worker/upload.go

```diff
@@ -0,0 +1,1126 @@
+			candidates = append(candidates, uploader)
+		}
+	}
+	mgr.mu.Unlock()
+
+	// return early if we have no queues left
+	if len(candidates) == 0 {
+		return nil
+	}
+
+loop:
+	for {
+		// grab ongoing slab uploads
+		upload.mu.Lock()
+		ongoing := upload.ongoing
+		upload.mu.Unlock()
+
+		// grab the slabs parents
+		var parents []slabID
+		for _, sID := range ongoing {
+			if sID == shard.sID {
+				break
+			}
+			parents = append(parents, sID)
+		}
+
+		// if this slab does not have more than 2 parents, we return the first
+		// (and thus best) candidate
+		if len(parents) < 3 {
+			return candidates[0]
+		}
+
+		// otherwise we wait, allowing the parents to complete
+		select {
+		case <-upload.doneShardTrigger:
+			continue loop
+		case <-shard.ctx.Done():
+			break loop
+		}
+	}
+
+	return nil
+}
+
+func (u *upload) finishSlabUpload(upload *slabUpload) {
+	// update ongoing slab history
+	u.mu.Lock()
+	for i, prev := range u.ongoing {
+		if prev == upload.sID {
+			u.ongoing = append(u.ongoing[:i], u.ongoing[i+1:]...)
+			break
+		}
+	}
+	u.mu.Unlock()
+
+	// cleanup contexts
+	upload.mu.Lock()
+	for _, shard := range upload.remaining {
+		shard.cancel()
+	}
+	upload.mu.Unlock()
+}
+
+func (u *upload) newSlabUpload(ctx context.Context, shards [][]byte) (*slabUpload, []*shardUpload, chan shardResp) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	// create slab id
+	var sID slabID
+	frand.Read(sID[:])
+	u.ongoing = append(u.ongoing, sID)
+
+	// create slab upload
+	slab := &slabUpload{
+		mgr: u.mgr,
+
+		upload:  u,
+		sID:     sID,
+		started: time.Now(),
+
+		overdriving: make(map[int]int, len(shards)),
+		remaining:   make(map[int]shardCtx, len(shards)),
+		sectors:     make([]object.Sector, len(shards)),
+	}
+
+	// prepare shard uploads
+	responseChan := make(chan shardResp)
+	uploads := make([]*shardUpload, len(shards))
+	for sI, shard := range shards {
+		// create the shard upload's cancel func
+		sCtx, cancel := context.WithCancel(ctx)
+		slab.remaining[sI] = shardCtx{ctx: sCtx, cancel: cancel}
+
+		// create the upload's span
+		sCtx, span := tracing.Tracer.Start(sCtx, "uploadShard")
+		span.SetAttributes(attribute.Bool("overdrive", false))
+		span.SetAttributes(attribute.Int("sector", sI))
+
+		// create the shard upload
+		uploads[sI] = &shardUpload{
+			upload:       u,
+			sID:          sID,
+			ctx:          sCtx,
+			responseChan: responseChan,
+
+			sector:      (*[rhpv2.SectorSize]byte)(shard),
+			sectorIndex: sI,
+		}
+	}
+
+	return slab, uploads, responseChan
+}
+
+func (u *upload) canUseUploader(ul *uploader, sID slabID) bool {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	_, excluded := u.excluded[ul.fcid]
+	if excluded {
+		return false
+	}
+
+	_, used := u.used[sID][ul.fcid]
+	return !used
+}
+
+func (u *upload) start(ctx context.Context, r io.Reader, rs api.RedundancySettings) chan slabResponse {
+	// create the response channel
+	slabsChan := make(chan slabResponse)
+
+	// launch a goroutine uploading the slabs
+	go func() {
```

Comment:
I think we can simplify this and get rid of a goroutine in the process.

The way I imagine `uploadManager.upload` on a high level is like this
```go
// some setup

for {
  case <- u.nextReadTrigger:
    readSlab() // can be blocking
    go uploadSlab() // basically what the inner goroutine in 'start' did
  case resp := <-slabsChan:
    if resp == nil {
      break // channel closed
    }
    responses = append(responses, resp) 
}

// sort responses, decorate etc.

```

That way it's more obvious to the reader that we don't just read slabs and upload but that there is a channel blocking reads. Also we save at least 2(?) goroutines. The one from `start` and one from `read`.

Reply:
I wanted to `for x := range chan` for once but this is better I guess. The exit condition is slightly awkward but probably fine. Have to break if you received enough responses.